File: aula29-generics/Generics.cs
using System;

interface I {
    object foo();
}
interface I<T> : I {
    T foo();
}
class A<T> : I<T> {
    public T foo(){
        return default(T);
    }
    object I.foo(){
        return "ola";
    }
}


public class Generics {
    public static void Main() {
        A<String> a = new A<String>();
        I<String> i1 = a;
        String m1 = a.foo();
        I i2 = a;
        Object o = i2.foo();
    }
}

File: aula29-generics/Generics.java

interface I<T> {
    T foo(T m);
}
class A<T> implements I<T> {
    public T foo(T m){
        return m;
    }
}


public class Generics {
    public static void main(String [] args) {
        A<String> a = new A<String>();
        I<String> i1 = a;
        I i2 = a;
        String m1 = i1.foo("ola");
        Object m2 = i2.foo("ola");
    }
}